### PR TITLE
feat: create new user

### PR DESCRIPTION
**Pull request by Vzart**
- added hashing method
- modified User and Department Configuration

### CODE REVIEW

File: src/Api/Controllers/ApiControllerBase.cs
Comment (StarryFolf) [inline src/Api/Controllers/ApiControllerBase.cs:1]: hahaha

File: src/Api/Controllers/UsersController.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+namespace Api.Controllers;
+
+[Route("api/v1/[controller]")]
+public class UsersController : ApiControllerBase
+{
+    [HttpPost]
+    public async Task<ActionResult<UserDTO>> CreateUser(CreateUserCommand command)
+    {
+        return await Mediator.Send(command);
```
Comment (ChienNQuang) [inline src/Api/Controllers/UsersController.cs:13]: Controller actions should return Result object, look into src/Application/Common/Models

File: src/Application/Helpers/SecurityUtil.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
+namespace Application.Helpers;
+
+using System.Security.Cryptography;
+using System.Text;
+
+public static class SecurityUtil
+{
+    public static string Hash(string input)
```
Comment (ChienNQuang) [inline src/Application/Helpers/SecurityUtil.cs:8]: Should salt this hash function

File: src/Application/Users/Commands/CreateUserCommand/CreateUserCommand.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+namespace Application.Users.Commands.CreateUserCommand;
+
+public record CreateUserCommand : IRequest<UserDTO>
+{
+    public string Username { get; init; }
+    public string Email { get; init; }
+    public string Password { get; set; }
+    public string FirstName { get; set; }
+    public string LastName { get; set; }
+    public Department Department { get; set; }
```
Comment (ChienNQuang) [inline src/Application/Users/Commands/CreateUserCommand/CreateUserCommand.cs:17]: Who can pass the whole department object in the http request?

File: src/Api/Controllers/UsersController.cs (hunk 1 above)
Comment (Vzart) [reply src/Api/Controllers/UsersController.cs:13]: Sorry, I just noticed

File: src/Application/Users/Queries/UserDTO.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+namespace Application.Users.Queries;
+
+public class UserDTO : IMapFrom<User>
+{
+    public string Username { get; set; }
+    public string Email { get; set; }
+    public string PasswordHash { get; set; }
+    public string FirstName { get; set; }
+    public string LastName { get; set; }
+    public Department Department { get; set; }
```
Comment (ChienNQuang) [inline src/Application/Users/Queries/UserDTO.cs:13]: Why UserDTO include an entity here?

File: src/Application/Users/Queries/UserDTO.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
+using Application.Common.Mappings;
+using Domain.Entities;
+
+namespace Application.Users.Queries;
+
+public class UserDTO : IMapFrom<User>
```
Comment (ChienNQuang) [inline src/Application/Users/Queries/UserDTO.cs:6]: Rename to UserDto to ensure PascalCasing

File: src/Application/Users/Queries/UserDTO.cs [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+{
+    public string Username { get; set; }
+    public string Email { get; set; }
+    public string PasswordHash { get; set; }
```
Comment (ChienNQuang) [inline src/Application/Users/Queries/UserDTO.cs:10]: No need to return PasswordHash here, who do it?

File: src/Infrastructure/ConfigureServices.cs [hunk 7]
```diff
@@ -52,6 +52,7 @@ private static IServiceCollection AddApplicationDbContext(this IServiceCollectio
     private static IServiceCollection RegisterRepositories(this IServiceCollection services)
     {
         services.AddScoped<IUnitOfWork, UnitOfWork>();
+        services.AddScoped<IUserRepository, UserRepository>();
```
Comment (ChienNQuang) [inline src/Infrastructure/ConfigureServices.cs:55]: Repositories should be registered before Unit Of Work

File: src/Application/Users/Commands/CreateUserCommand/CreateUserCommand.cs (hunk 3 above)
Comment (Vzart) [reply src/Application/Users/Commands/CreateUserCommand/CreateUserCommand.cs:17]: ah yeah I c, understandable

File: src/Application/Users/Queries/UserDTO.cs (hunk 5 above)
Comment (Vzart) [reply src/Application/Users/Queries/UserDTO.cs:6]: okay

File: src/Infrastructure/Repositories/UnitOfWork.cs [hunk 8]
```diff
@@ -12,20 +12,22 @@ public class UnitOfWork : IUnitOfWork
...
 
-    public UnitOfWork(IDbTransaction dbTransaction, IMediator mediator, DbContext context)
+    public UnitOfWork(IDbTransaction dbTransaction, IMediator mediator, DbContext context, IUserRepository userRepository)
     {
         // Baseline
         _dbTransaction = dbTransaction;
         _mediator = mediator;
         _context = context;
-
+        UserRepository = userRepository;
```
Comment (ChienNQuang) [inline src/Infrastructure/Repositories/UnitOfWork.cs:21]: ????

File: src/Infrastructure/Repositories/UserRepository.cs [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+    public UserRepository(IDbConnection connection, IDbTransaction transaction)
+    {
+        _connection = connection;
+        _transaction = transaction;
+    }
+    
+    public async Task<User> CreateUserAsync(User user)
+    {
+        var commandText = $"INSERT INTO " +
```
Comment (ChienNQuang) [inline src/Infrastructure/Repositories/UserRepository.cs:21]: name it sql is more than enough

File: src/Infrastructure/Repositories/UserRepository.cs [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+                          "User(username, email, password_hash, firstName, " +
+                          "lastName, department_id, role, position, isActive) " +
+                          "VALUES (@username, @email, @password_hash, @firstName," +
+                          "@lastname, @department_id, @role, @position, @isActive)";
+
+        var queryArguments = new
```
Comment (ChienNQuang) [inline src/Infrastructure/Repositories/UserRepository.cs:27]: you can pass an object with the property order as the schema and it is still ok

File: src/Infrastructure/Repositories/UserRepository.cs [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+            firstName = user.FirstName,
+            lastName = user.LastName,
+            department_id = user.Department.Id,
+            role = user.Role,
+            position = user.Position,
+            isActive = user.IsActive
+        };
+
+        var affectedRows = await _connection.ExecuteAsync(commandText, queryArguments, transaction: _transaction);
+        if (affectedRows > 0)
```
Comment (ChienNQuang) [inline src/Infrastructure/Repositories/UserRepository.cs:41]: affectedRows is not needed, transaction is not commited

**Review by ChienNQuang — CHANGES_REQUESTED** [review]
Look for all changes and re do them

File: src/Application/Helpers/SecurityUtil.cs (hunk 2 above)
Comment (kaitoz11) [reply src/Application/Helpers/SecurityUtil.cs:8]: And salt should be an environment variable.

File: src/Infrastructure/Repositories/UserRepository.cs (hunk 9 above)
Comment (kaitoz11) [reply src/Infrastructure/Repositories/UserRepository.cs:21]: > name it sql is more than enough

Using `commandText` to differentiate between a "Command" and a "Query"

**Review by kaitoz11 — COMMENTED** [review]
Well... You better satisfy @ChienNQuang

File: src/Infrastructure/Persistence/Configurations/DepartmentConfiguration.cs [hunk 12]
```diff
@@ -9,6 +10,8 @@ public class DepartmentConfiguration : IEntityTypeConfiguration<Department>
     public void Configure(EntityTypeBuilder<Department> builder)
     {
         builder.HasKey(x => x.Id);
+        builder.Property(x => x.Id)
+            .HasValueGenerator<GuidValueGenerator>();
```
Comment (ChienNQuang) [inline src/Infrastructure/Persistence/Configurations/DepartmentConfiguration.cs:14]: should use methods other than this, as for https://github.com/dotnet/efcore/issues/28024

File: src/Infrastructure/Persistence/Configurations/UserConfiguration.cs [hunk 13]
```diff
@@ -9,6 +10,8 @@ public class UserConfiguration : IEntityTypeConfiguration<User>
     public void Configure(EntityTypeBuilder<User> builder)
     {
         builder.HasKey(x => x.Id);
+        builder.Property(x => x.Id)
+            .HasValueGenerator<GuidValueGenerator>();
```
Comment (ChienNQuang) [inline src/Infrastructure/Persistence/Configurations/UserConfiguration.cs:14]: Same reason with DepartmentConfiguration

**Review by ChienNQuang — CHANGES_REQUESTED** [review]
Need change many things

File: src/Infrastructure/Repositories/UserRepository.cs (hunk 9 above)
Comment (ChienNQuang) [reply src/Infrastructure/Repositories/UserRepository.cs:21]: This is at Data/Persistence level, the difference between a command and query should not be a matter here

File: src/Application/Users/Queries/UserDTO.cs (hunk 5 above)
Comment (ChienNQuang) [reply src/Application/Users/Queries/UserDTO.cs:6]: also move this to src/Application/Common/Models/Dtos

**Review by StarryFolf — APPROVED** [review]
lgtm

File: src/Api/Controllers/UsersController.cs [hunk 14]
```diff
@@ -8,8 +9,9 @@ namespace Api.Controllers;
 public class UsersController : ApiControllerBase
 {
     [HttpPost]
-    public async Task<ActionResult<UserDTO>> CreateUser(CreateUserCommand command)
+    public async Task<ActionResult<Result<UserDto>>> CreateUser(CreateUserCommand command)
     {
-        return await Mediator.Send(command);
+        var result = await Mediator.Send(command);
+        return Created("", Result<UserDto>.Succeed(result));
```
Comment (ChienNQuang) [inline src/Api/Controllers/UsersController.cs:15]: Should use either Ok or pointing to an actual location

File: src/Application/Users/Commands/CreateUserCommand/CreateUserCommand.cs [hunk 15]
```diff
@@ -38,14 +37,15 @@ public async Task<UserDTO> Handle(CreateUserCommand request, CancellationToken c
             Email = request.Email,
             FirstName = request.FirstName,
             LastName = request.LastName,
-            Department = request.Department,
+            Department = await _uow.DepartmentRepository.GetByIdAsync(request.DepartmentId),
```
Comment (ChienNQuang) [inline src/Application/Users/Commands/CreateUserCommand/CreateUserCommand.cs:40]: split this to a separate statement, what if the department is null?

File: src/Infrastructure/Repositories/DepartmentRepository.cs [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+using Domain.Entities;
+
+namespace Infrastructure.Repositories;
+
+public class DepartmentRepository : IDepartmentRepository
+{
+    private readonly IDbConnection _connection;
+    private readonly IDbTransaction _transaction;
+
+    public DepartmentRepository(IDbConnection connection, IDbTransaction transaction)
```
Comment (ChienNQuang) [inline src/Infrastructure/Repositories/DepartmentRepository.cs:13]: what's the point of injecting without assigning it? and is it needed here now?

File: src/Infrastructure/Repositories/DepartmentRepository.cs [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+    {
+        _connection = connection;
+    }
+
+    public async Task<Department> GetByIdAsync(Guid id)
+    {
+        var sql = @"SELECT id, name FROM Department WHERE id = @id";
+        return await _connection.QuerySingleOrDefault<Task<Department>>(sql, new { id });
```
Comment (ChienNQuang) [inline src/Infrastructure/Repositories/DepartmentRepository.cs:21]: The result of that method is never a Task

File: src/Infrastructure/Repositories/DepartmentRepository.cs (hunk 16 above)
Comment (Vzart) [reply src/Infrastructure/Repositories/DepartmentRepository.cs:13]: I resolve it in the latest commit

**Review by ChienNQuang — CHANGES_REQUESTED** [review]
A little more resolving and it should be fine

File: src/Api/Controllers/UsersController.cs (hunk 14 above)
Comment (Vzart) [reply src/Api/Controllers/UsersController.cs:15]: isn't that create should return 201 ?

File: src/Infrastructure/Repositories/DepartmentRepository.cs (hunk 16 above)
Comment (ChienNQuang) [reply src/Infrastructure/Repositories/DepartmentRepository.cs:13]: ok

File: src/Api/Controllers/UsersController.cs (hunk 14 above)
Comment (ChienNQuang) [reply src/Api/Controllers/UsersController.cs:15]: leaving a blank in Created() means nothing, if you insist on using it then adhere to the full HTTP semantics by pointing to the real location that created resource lie in
If you cannot point to one, leave it as Ok is sufficient

File: src/Infrastructure/Repositories/DepartmentRepository.cs (hunk 17 above)
Comment (Vzart) [reply src/Infrastructure/Repositories/DepartmentRepository.cs:21]: what do you mean here, it a Taskl<Department>, we use async from top to bottom no ?

File: src/Api/Controllers/UsersController.cs (hunk 14 above)
Comment (Vzart) [reply src/Api/Controllers/UsersController.cs:15]: noted , thanks

File: src/Application/Users/Commands/CreateUserCommand/CreateUserCommand.cs [hunk 18]
```diff
@@ -30,14 +30,19 @@ public CreateUserCommandHandler(IUnitOfWork uow, IMapper mapper)
     }
     public async Task<UserDto> Handle(CreateUserCommand request, CancellationToken cancellationToken)
     {
+        var department = await _uow.DepartmentRepository.GetByIdAsync(request.DepartmentId);
+        if (department is null)
+        {
+            throw new KeyNotFoundException($"department {request.DepartmentId} doesn't existed");
```
Comment (ChienNQuang) [inline src/Application/Users/Commands/CreateUserCommand/CreateUserCommand.cs:36]: 'doesn't existed'?

**Review by ChienNQuang — CHANGES_REQUESTED** [review]
Nearly fine, but where is the hash function?

**Review by ChienNQuang — CHANGES_REQUESTED** [review]
Nearly fine, but where is the hash function?

**Review by ChienNQuang — APPROVED** [review]

**Review by ChienNQuang — APPROVED** [review]

File: src/Application/Common/Models/Dtos/UserDto.cs [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+public class UserDto : IMapFrom<User>
+{
+    public string Username { get; set; }
+    public string Email { get; set; }
+    public string FirstName { get; set; }
+    public string LastName { get; set; }
+    public Guid DepartmentId { get; set; }
+    public string Role { get; set; }
+    public string Position { get; set; }
+    public bool IsActive { get; set; }
```
Comment (kaitoz11) [inline src/Application/Common/Models/Dtos/UserDto.cs:16]: this should be `IsActivated`!

**Review by kaitoz11 — APPROVED** [review]

File: src/Application/Common/Models/Dtos/UserDto.cs (hunk 19 above)
Comment (Vzart) [reply src/Application/Common/Models/Dtos/UserDto.cs:16]: IsActive is different from IsActivated according to @ChienNQuang

Comment (kaitoz11) [reply src/Application/Common/Models/Dtos/UserDto.cs:16]: @ChienNQuang can you explain?

**Pull request merged**